File: paradigms/object-oriented.js
/**
 * Object-Oriented Programming - JavaScript
 * 
 * Programming paradigm based on classes, objects, inheritance, polymorphism, and encapsulation.
 */

// Class definition
class Animal {
    constructor(name, species) {
        this.name = name;
        this.species = species;
    }
    
    // Method
    speak() {
        return `${this.name} makes a sound`;
    }
    
    // Getter
    getInfo() {
        return `${this.name} is a ${this.species}`;
    }
}

// Inheritance
class Dog extends Animal {
    constructor(name, breed) {
        super(name, 'Dog');
        this.breed = breed;
    }
    
    // Method overriding (polymorphism)
    speak() {
        return `${this.name} barks: Woof!`;
    }
    
    // Additional method
    fetch() {
        return `${this.name} fetches the ball`;
    }
}

class Cat extends Animal {
    constructor(name, color) {
        super(name, 'Cat');
        this.color = color;
    }
    
    speak() {
        return `${this.name} meows: Meow!`;
    }
    
    purr() {
        return `${this.name} purrs contentedly`;
    }
}

// Encapsulation with private fields (ES2022)
class BankAccount {
    #balance = 0; // Private field
    
    constructor(owner) {
        this.owner = owner;
    }
    
    deposit(amount) {
        if (amount > 0) {
            this.#balance += amount;
            return `Deposited $${amount}. New balance: $${this.#balance}`;
        }
        return 'Invalid deposit amount';
    }
    
    withdraw(amount) {
        if (amount > 0 && amount <= this.#balance) {
            this.#balance -= amount;
            return `Withdrew $${amount}. New balance: $${this.#balance}`;
        }
        return 'Insufficient funds or invalid amount';
    }
    
    getBalance() {
        return this.#balance;
    }
}

// Abstraction - Interface-like pattern
class Shape {
    area() {
        throw new Error('Method must be implemented');
    }
    
    perimeter() {
        throw new Error('Method must be implemented');
    }
}

class Rectangle extends Shape {
    constructor(width, height) {
        super();
        this.width = width;
        this.height = height;
    }
    
    area() {
        return this.width * this.height;
    }
    
    perimeter() {
        return 2 * (this.width + this.height);
    }
}

class Circle extends Shape {
    constructor(radius) {
        super();
        this.radius = radius;
    }
    
    area() {
        return Math.PI * this.radius ** 2;
    }
    
    perimeter() {
        return 2 * Math.PI * this.radius;
    }
}

// Example usage
console.log('=== Object-Oriented Programming ===\n');

// Inheritance and Polymorphism
const dog = new Dog('Buddy', 'Golden Retriever');
const cat = new Cat('Whiskers', 'Orange');

console.log('Inheritance & Polymorphism:');
console.log(dog.getInfo());
console.log(dog.speak());
console.log(cat.getInfo());
console.log(cat.speak());

console.log('\nEncapsulation:');
const account = new BankAccount('John Doe');
console.log(account.deposit(100));
console.log(account.deposit(50));
console.log(account.withdraw(30));
console.log(`Balance: $${account.getBalance()}`);

console.log('\nAbstraction:');
const rectangle = new Rectangle(5, 10);
const circle = new Circle(5);
console.log(`Rectangle area: ${rectangle.area()}, perimeter: ${rectangle.perimeter()}`);
console.log(`Circle area: ${circle.area().toFixed(2)}, perimeter: ${circle.perimeter().toFixed(2)}`);

module.exports = { Animal, Dog, Cat, BankAccount, Shape, Rectangle, Circle };

File: paradigms/object-oriented.py
"""
Object-Oriented Programming - Python

Programming paradigm based on classes, objects, inheritance, polymorphism, and encapsulation.
"""

from abc import ABC, abstractmethod


# Class definition
class Animal:
    def __init__(self, name, species):
        self.name = name
        self.species = species
    
    def speak(self):
        return f"{self.name} makes a sound"
    
    def get_info(self):
        return f"{self.name} is a {self.species}"


# Inheritance
class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name, 'Dog')
        self.breed = breed
    
    # Method overriding (polymorphism)
    def speak(self):
        return f"{self.name} barks: Woof!"
    
    # Additional method
    def fetch(self):
        return f"{self.name} fetches the ball"


class Cat(Animal):
    def __init__(self, name, color):
        super().__init__(name, 'Cat')
        self.color = color
    
    def speak(self):
        return f"{self.name} meows: Meow!"
    
    def purr(self):
        return f"{self.name} purrs contentedly"


# Encapsulation with private attributes
class BankAccount:
    def __init__(self, owner):
        self.owner = owner
        self._balance = 0  # Protected attribute
    
    def deposit(self, amount):
        if amount > 0:
            self._balance += amount
            return f"Deposited ${amount}. New balance: ${self._balance}"
        return "Invalid deposit amount"
    
    def withdraw(self, amount):
        if amount > 0 and amount <= self._balance:
            self._balance -= amount
            return f"Withdrew ${amount}. New balance: ${self._balance}"
        return "Insufficient funds or invalid amount"
    
    def get_balance(self):
        return self._balance


# Abstraction - Abstract base class
class Shape(ABC):
    @abstractmethod
    def area(self):
        pass
    
    @abstractmethod
    def perimeter(self):
        pass


class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def area(self):
        return self.width * self.height
    
    def perimeter(self):
        return 2 * (self.width + self.height)


class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius
    
    def area(self):
        return 3.14159 * self.radius ** 2
    
    def perimeter(self):
        return 2 * 3.14159 * self.radius


# Example usage
if __name__ == "__main__":
    print("=== Object-Oriented Programming ===\n")
    
    # Inheritance and Polymorphism
    dog = Dog('Buddy', 'Golden Retriever')
    cat = Cat('Whiskers', 'Orange')
    
    print("Inheritance & Polymorphism:")
    print(dog.get_info())
    print(dog.speak())
    print(cat.get_info())
    print(cat.speak())
    
    print("\nEncapsulation:")
    account = BankAccount('John Doe')
    print(account.deposit(100))
    print(account.deposit(50))
    print(account.withdraw(30))
    print(f"Balance: ${account.get_balance()}")
    
    print("\nAbstraction:")
    rectangle = Rectangle(5, 10)
    circle = Circle(5)
    print(f"Rectangle area: {rectangle.area()}, perimeter: {rectangle.perimeter()}")
    print(f"Circle area: {circle.area():.2f}, perimeter: {circle.perimeter():.2f}")

